cpp/src_prims/stats/minmax.h

```diff
@@ -166,16 +166,18 @@ __global__ void minmaxKernel(const T* data, const unsigned int* rowids,
  * @param globalmax final col-wise global maximum (size = ncols)
  * @param sampledcols output sampled data. Pass nullptr if you don't need this
  * @param init_val initial minimum value to be
+ * @param sharedMemPerBlk max shared mem available per block. This is being
+ * explicitly requested since querying it inside prims is not good for perf.
  * @param stream: cuda stream
  * @note This method makes the following assumptions:
  * 1. input and output matrices are assumed to be col-major
  * 2. ncols is small enough to fit the whole of min/max values across all cols
  *    in shared memory
  */
 template <typename T, int TPB = 512>
-void minmax(const T* data, const unsigned int* rowids,
-            const unsigned int* colids, int nrows, int ncols, int row_stride,
-            T* globalmin, T* globalmax, T* sampledcols, cudaStream_t stream) {
+void minmax(const T* data, const unsigned* rowids, const unsigned* colids,
+            int nrows, int ncols, int row_stride, T* globalmin, T* globalmax,
```

Comment:
Just thinking out loud, maybe way off base... I worry a bit about requiring the caller to pass such device-specific info in a low-level math function. Maybe the ideal would be to have a function `GetDevicePropertiesCached` that does its own caching (in a thread-local static) rather than putting it in the handle, which requires us to pass it around?

Reply:
If we have reasonable defaults and a default parameter, that would be another approach that helps... like for best performance the caller needs to pass in the device info, but it will work even without that.

Reply:
SHMEM size varies all across the board. So hard to get a reasonable estimate here.

Reply:
Agree with Vishal. It's impossible to have a reasonable default, without compromising on perf.

Another point is that the moment we introduce caching, we'll have some form a state introduced into the ml-prims, cuml world, which is clearly against our dev-guide. The only place in cuML which has state in our library has been `cumlHandle`.

Reply:
Yeah, I think it looks better this way. Thanks!

Reply:
Agreed that in the long run exposing the cumlHandle into prims seems appealing.